File: 0684/0684.py
class Solution:
    def findRedundantConnection1(self, edges: List[List[int]]) -> List[int]:
        # Union find O(nlog*n) = O(n)
        def findRoot(x, tree):
            if tree[x] == -1: return x
            root = findRoot(tree[x], tree)
            tree[x] = root
            return root
        
        tree = [-1]*(len(edges)+1)
        for e in edges:
            a = findRoot(e[0], tree)
            b = findRoot(e[1], tree)
            if a == b:
                return e
            tree[a] = b
            
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        # dfs O(n^2)
        def dfs(a, b):
            visited = set()
            stack = [a]
            while stack:
                node = stack.pop()
                if node == b:
                    return True
                visited.add(node)
                for neighbor in graph[node]:
                    if neighbor not in visited:
                        stack.append(neighbor)
            return False
        
        graph = collections.defaultdict(list)
        res = []
        for a,b in edges:
            if a in graph and b in graph:
                if dfs(a, b):
                    res = [a, b]
            graph[a].append(b)
            graph[b].append(a)

        return res

File: notes/0684/0684.cpp
// class Solution {
// public:
//     vector<int> findRedundantConnection(vector<vector<int>>& edges) {
//         const int N = edges.size();
//         vector<int> tree(N+1, -1);
        
//         for (auto p: edges) {
//             int root_a = findRoot(p[0], tree);
//             int root_b = findRoot(p[1], tree);
//             if (root_a == root_b) return p;
//             tree[root_b] = root_a;
//         }
//         return {};
//     }
// private:
//     int findRoot(int x, vector<int>& tree) {
//         if (tree[x]==-1) return x;
//         int root = findRoot(tree[x], tree);
//         tree[x] = root;
//         return root;
//     }
// };
class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        unordered_map<int, unordered_set<int>> m;
        for (auto e: edges) {
            if (m.count(e[0]) && m.count(e[1])) {
                if (bfs(e[0], e[1], m)) return e;
            }
            m[e[0]].insert(e[1]);
            m[e[1]].insert(e[0]);
        }
        return {};
    }
private:
    bool bfs(int a, int b, unordered_map<int, unordered_set<int>>& m) {
        queue<int> q;
        unordered_set<int> visited;
        
        q.push(a);
        visited.insert(a);
        while (!q.empty()) {
            int cur = q.front(); q.pop();
            if (cur == b) return true;
            for (auto nei: m[cur]) {
                if (!visited.count(nei)) {
                    visited.insert(nei);
                    q.push(nei);
                }
            }
        }
        return false;
    }
};
